Run distributed_test with NCCL_ASYNC_ERROR_HANDLING

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52887 [Dist Debugability] Refactor and fix DDP model check during init
* **#52619 Run distributed_test with NCCL_ASYNC_ERROR_HANDLING**

Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

rohan-varma committed:
Run distributed_test with NCCL_ASYNC_ERROR_HANDLING

Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -319,7 +319,7 @@ class TestDistBackend(MultiProcessTestCase):
         os.environ["MASTER_ADDR"] = str(MASTER_ADDR)
         os.environ["MASTER_PORT"] = str(MASTER_PORT)
-        # os.environ["WORLD_SIZE"] = str(WORLD_SIZE)
+        os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "1"
```
We should add a comment here that NCCL_BLOCKING_WAIT overrides this and as a result tests like `test_batch_isend_irecv_nccl` would be fine.

Btw are there no tests that currently enable `NCCL_ASYNC_ERROR_HANDLING`? I couldn't find any test files enabling this. @osalpekar Should we add some tests for that verify `NCCL_ASYNC_ERROR_HANDLING` behavior?

rohan-varma replied:
There's no unit tests AFAIK, but I think there is an internal mechanism for testing that it works appropriately

osalpekar replied:
Sure I can add some unit tests for NCCL Async Error Handling. We have a pretty useful flow for verifying correctness which is more of an integration test since it tests torchelastic + Async Error Handling e2e, but a unit tests would be good to have as well.

rohan-varma committed:
Update on "Run distributed_test with NCCL_ASYNC_ERROR_HANDLING"


Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Run distributed_test with NCCL_ASYNC_ERROR_HANDLING"


Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Should we do this for `test_c10d` tests as well?

rohan-varma committed:
Update on "Run distributed_test with NCCL_ASYNC_ERROR_HANDLING"


Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Run distributed_test with NCCL_ASYNC_ERROR_HANDLING"


Runs this test suite with nccl_async_error_handling enabled. It is the
default to run many distributed training jobs, and can also help catch
errors/hangs in tests more easily. We don't expect any changes in the actual
existing tests since they shouldn't have any hangs.

Also removes a commented out line

Differential Revision: [D26588108](https://our.internmc.facebook.com/intern/diff/D26588108/)

[ghstack-poisoned]

rohan-varma commented:
Failures are due to unrelated mypy errors
